ethcore/src/block.rs

```diff
@@ -405,14 +404,22 @@ impl LockedBlock {
 	/// Provide a valid seal in order to turn this into a `SealedBlock`.
 	///
 	/// NOTE: This does not check the validity of `seal` with the engine.
-	pub fn seal(self, engine: &EthEngine, seal: Vec<Bytes>) -> Result<SealedBlock, BlockError> {
-		let expected_seal_fields = engine.seal_fields(&self.block.header);
+	pub fn seal(self, engine: &EthEngine, seal: Vec<Bytes>) -> Result<SealedBlock, Error> {
+		let expected_seal_fields = engine.seal_fields(&self.header);
 		let mut s = self;
 		if seal.len() != expected_seal_fields {
-			return Err(BlockError::InvalidSealArity(
-				Mismatch { expected: expected_seal_fields, found: seal.len() }));
+			Err(BlockError::InvalidSealArity(Mismatch {
+				expected: expected_seal_fields,
+				found: seal.len()
+			}))?;
 		}
+
 		s.block.header.set_seal(seal);
+
+		if let Some(new_header) = engine.on_seal_block(&s.block)? {
```

Comment:
Maybe make `Engine::on_seal_block` take a `&mut` of the block instead. This avoids a copy.

Reply:
https://github.com/goerli/parity-goerli/pull/71